File: analysis/rap-0-R/data-collection/generate-synthetic-data.R
#' ---
#' title: "Generate Synthetic Edmonton Cafe Data (R Version)"
#' author: "Research Team"
#' date: "Last Updated: `r Sys.Date()`"
#' ---

#' Generate synthetic Edmonton cafe data for RAP-0-R testing
#' 
#' This script creates a realistic SQLite database with synthetic cafe data
#' for Square One Coffee's competition analysis pipeline.
#' 
#' Data includes:
#' - Cafe names, locations, and contact info
#' - Operating hours and menu offerings
#' - Pricing data and cafe characteristics
#' - Geographic coordinates (Edmonton area)

# ---- load-packages -----------------------------------------------------------
library(DBI)
library(RSQLite)
library(dplyr)
library(tidyr)
library(stringr)

# ---- declare-globals ---------------------------------------------------------
# Ensure data-private/raw directory exists
output_dir <- "../../../data-private/raw"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# Database path
db_path <- file.path(output_dir, "edmonton_cafes.sqlite")

# Remove existing database if it exists
if (file.exists(db_path)) {
  file.remove(db_path)
  message("Removed existing database")
}

# Edmonton neighborhoods
neighborhoods <- c(
  "Downtown", "Oliver", "Garneau", "Whyte Avenue", "Bonnie Doon",
  "Westmount", "Old Strathcona", "Ritchie", "Highlands", "Jasper Avenue",
  "Alberta Avenue", "124 Street", "Capilano", "Belgravia", "Riverdale"
)

# Edmonton lat/long boundaries (approximate)
lat_bounds <- c(min = 53.45, max = 53.62)
lng_bounds <- c(min = -113.65, max = -113.40)

# ---- generate-data -----------------------------------------------------------
cat("\n## Generating Synthetic Data\n\n")

# Square One Coffee locations
soc_locations <- paste0("Square One Coffee - ", 
                       c("Oliver", "Downtown", "Whyte Avenue", 
                         "Westmount", "124 Street", "Ritchie"))

# Generate competitor names
set.seed(42)  # For reproducibility
prefixes <- c("The", "Cafe", "Coffee", "Brew", "Bean", "Roast", "Morning", "Daily")
middles <- c("Central", "House", "Bar", "Shop", "Co", "Collective", "Studio", "Lab")
suffixes <- c("Cafe", "Coffee", "Roasters", "Co.", "House", "Bar", "Kitchen")

competitor_cafes <- character(24)
for (i in 1:24) {
  if (runif(1) < 0.3) {
    competitor_cafes[i] <- paste(sample(prefixes, 1), sample(middles, 1))
  } else {
    competitor_cafes[i] <- paste(sample(prefixes, 1), sample(middles, 1), sample(suffixes, 1))
  }
}

all_cafes <- c(soc_locations, competitor_cafes)

# Cafe characteristics
cafe_types <- c("specialty_coffee", "espresso_bar", "full_service_cafe", "coffee_shop", "roastery_cafe")
ownerships <- c("independent", "small_chain", "regional_chain", "national_chain")
ambiances <- c("modern_minimalist", "cozy_traditional", "industrial_chic", "community_hub", "grab_and_go")
parking_options <- c("street_only", "nearby_lot", "dedicated_parking", "no_parking")
food_offerings <- c("pastries_only", "sandwiches_pastries", "full_menu", "none")

# ---- create-cafe-data --------------------------------------------------------
cat("Creating cafe records...\n")

cafes_data <- tibble(
  name = all_cafes
) %>%
  mutate(
    is_soc = str_detect(name, regex("square one", ignore_case = TRUE)),
    
    # SOC cafes have consistent characteristics
    cafe_type = if_else(is_soc, "specialty_coffee", sample(cafe_types, n(), replace = TRUE)),
    ownership = if_else(is_soc, "independent", sample(ownerships, n(), replace = TRUE)),
    avg_beverage_price = if_else(is_soc, 
                                  runif(n(), 4.50, 6.00),
                                  runif(n(), 3.00, 7.50)),
    has_food = if_else(is_soc, 
                      "sandwiches_pastries",
                      sample(food_offerings, n(), replace = TRUE)),
    has_wifi = if_else(is_soc, "yes", sample(c("yes", "no", "limited"), n(), replace = TRUE)),
    seating_capacity = if_else(is_soc, 
                               as.integer(runif(n(), 20, 45)),
                               as.integer(runif(n(), 10, 60))),
    ambiance = if_else(is_soc, "modern_minimalist", sample(ambiances, n(), replace = TRUE)),
    google_rating = if_else(is_soc, 
                           runif(n(), 4.3, 4.8),
                           runif(n(), 3.5, 4.9)),
    review_count = if_else(is_soc, 
                          as.integer(runif(n(), 150, 500)),
                          as.integer(runif(n(), 20, 400))),
    
    # Assign neighborhood
    neighborhood = neighborhoods[(row_number() - 1) %% length(neighborhoods) + 1],
    
    # Generate coordinates
    latitude = runif(n(), lat_bounds["min"], lat_bounds["max"]),
    longitude = runif(n(), lng_bounds["min"], lng_bounds["max"]),
    
    # Generate address
    street_num = sample(100:9999, n()),
    street_name = sample(c("Jasper Ave", "Whyte Ave", "124 St", "104 St", "82 Ave", 
                          "Gateway Blvd", "Calgary Trail"), n(), replace = TRUE),
    address = paste0(street_num, " ", street_name, ", Edmonton, AB"),
    
    # Contact info
    phone = sprintf("780-%03d-%04d", sample(100:999, n()), sample(1000:9999, n())),
    website = if_else(runif(n()) > 0.3, 
                     paste0("https://", str_replace_all(tolower(substr(name, 1, 20)), " ", ""), ".com"),
                     NA_character_),
    
    # Hours
    hours_weekday = if_else(runif(n()) > 0.3, "7:00 AM - 6:00 PM", "6:30 AM - 7:00 PM"),
    hours_weekend = if_else(runif(n()) > 0.3, "8:00 AM - 5:00 PM", "8:00 AM - 6:00 PM"),
    
    # Opening date
    date_opened = as.character(as.Date(sprintf("%d-%02d-01", 
                                              sample(2010:2024, n(), replace = TRUE),
                                              sample(1:12, n(), replace = TRUE)))),
    
    # Instagram
    instagram_handle = if_else(runif(n()) > 0.2,
                              paste0("@", str_replace_all(tolower(substr(name, 1, 20)), " ", "")),
                              NA_character_),
    
    # Parking
    parking_availability = sample(parking_options, n(), replace = TRUE),
    
    # Timestamps
    created_at = as.character(Sys.time()),
    updated_at = as.character(Sys.time())
  ) %>%
  select(-is_soc, -street_num, -street_name) %>%
  # Round numeric columns
  mutate(
    avg_beverage_price = round(avg_beverage_price, 2),
    google_rating = round(google_rating, 1),
    latitude = round(latitude, 6),
    longitude = round(longitude, 6)
  )

cat(sprintf("✓ Generated %d cafe records\n", nrow(cafes_data)))

# ---- write-to-database -------------------------------------------------------
cat("\n## Writing to Database\n\n")

# Connect to database
conn <- dbConnect(RSQLite::SQLite(), db_path)

# Write table
dbWriteTable(conn, "cafes", cafes_data, overwrite = TRUE)

# Verify
total_cafes <- dbGetQuery(conn, "SELECT COUNT(*) as count FROM cafes")$count
soc_cafes <- dbGetQuery(conn, "SELECT COUNT(*) as count FROM cafes WHERE name LIKE '%Square One%'")$count

cat("\n✅ Synthetic data generated successfully!\n")
cat(sprintf("   Database: %s\n", db_path))
cat(sprintf("   Total cafes: %d\n", total_cafes))
cat(sprintf("   - Square One Coffee locations: %d\n", soc_cafes))
cat(sprintf("   - Competitor cafes: %d\n", total_cafes - soc_cafes))
cat(sprintf("   Neighborhoods covered: %d\n", length(neighborhoods)))

# Close connection
dbDisconnect(conn)

cat("\n✓ Database created and verified\n")

File: analysis/rap-1/data-collection/generate-synthetic-data.py
#!/usr/bin/env python3
"""
Generate synthetic Edmonton cafe data for RAP-1 testing

This script creates a realistic SQLite database with synthetic cafe data
for Square One Coffee's competition analysis pipeline.

Data includes:
- Cafe names, locations, and contact info
- Operating hours and menu offerings
- Pricing data and cafe characteristics
- Geographic coordinates (Edmonton area)
"""

import sqlite3
import random
from datetime import datetime
import os

# Ensure data-private/raw directory exists
os.makedirs("../../../data-private/raw", exist_ok=True)

# Database path
db_path = "../../../data-private/raw/edmonton_cafes.sqlite"

# Connect to database
conn = sqlite3.connect(db_path)
cursor = conn.cursor()

# Create cafes table
cursor.execute("""
CREATE TABLE IF NOT EXISTS cafes (
    cafe_id INTEGER PRIMARY KEY AUTOINCREMENT,
    name TEXT NOT NULL,
    address TEXT,
    neighborhood TEXT,
    latitude REAL,
    longitude REAL,
    phone TEXT,
    website TEXT,
    cafe_type TEXT,
    ownership TEXT,
    avg_beverage_price REAL,
    has_food TEXT,
    has_wifi TEXT,
    seating_capacity INTEGER,
    ambiance TEXT,
    parking_availability TEXT,
    hours_weekday TEXT,
    hours_weekend TEXT,
    date_opened TEXT,
    instagram_handle TEXT,
    google_rating REAL,
    review_count INTEGER,
    created_at TEXT,
    updated_at TEXT
)
""")

# Edmonton neighborhoods
neighborhoods = [
    "Downtown", "Oliver", "Garneau", "Whyte Avenue", "Bonnie Doon",
    "Westmount", "Old Strathcona", "Ritchie", "Highlands", "Jasper Avenue",
    "Alberta Avenue", "124 Street", "Capilano", "Belgravia", "Riverdale"
]

# Cafe name components for variety
prefixes = ["The", "Cafe", "Coffee", "Brew", "Bean", "Roast", "Morning", "Daily"]
middles = ["Central", "House", "Bar", "Shop", "Co", "Collective", "Studio", "Lab"]
suffixes = ["Cafe", "Coffee", "Roasters", "Co.", "House", "Bar", "Kitchen"]

# Generate cafe names (including Square One locations)
soc_locations = [
    "Square One Coffee - Oliver",
    "Square One Coffee - Downtown",
    "Square One Coffee - Whyte Avenue",
    "Square One Coffee - Westmount",
    "Square One Coffee - 124 Street",
    "Square One Coffee - Ritchie"
]

# Generate additional competitor cafes
competitor_cafes = []
for i in range(24):  # 24 competitors + 6 SOC = 30 total cafes
    if random.random() < 0.3:
        name = f"{random.choice(prefixes)} {random.choice(middles)}"
    else:
        name = f"{random.choice(prefixes)} {random.choice(middles)} {random.choice(suffixes)}"
    competitor_cafes.append(name)

all_cafes = soc_locations + competitor_cafes

# Edmonton lat/long boundaries (approximate)
lat_min, lat_max = 53.45, 53.62
lng_min, lng_max = -113.65, -113.40

# Cafe types
cafe_types = ["specialty_coffee", "espresso_bar", "full_service_cafe", "coffee_shop", "roastery_cafe"]
ownerships = ["independent", "small_chain", "regional_chain", "national_chain"]
ambiances = ["modern_minimalist", "cozy_traditional", "industrial_chic", "community_hub", "grab_and_go"]
parking_options = ["street_only", "nearby_lot", "dedicated_parking", "no_parking"]

# Insert synthetic cafe data
cafes_data = []
for i, cafe_name in enumerate(all_cafes):
    is_soc = "Square One" in cafe_name
    
    # SOC cafes have consistent characteristics
    if is_soc:
        cafe_type = "specialty_coffee"
        ownership = "independent"
        avg_price = round(random.uniform(4.50, 6.00), 2)
        has_food = "sandwiches_pastries"
        has_wifi = "yes"
        seating = random.randint(20, 45)
        ambiance = "modern_minimalist"
        google_rating = round(random.uniform(4.3, 4.8), 1)
        review_count = random.randint(150, 500)
    else:
        cafe_type = random.choice(cafe_types)
        ownership = random.choice(ownerships)
        avg_price = round(random.uniform(3.00, 7.50), 2)
        has_food = random.choice(["pastries_only", "sandwiches_pastries", "full_menu", "none"])
        has_wifi = random.choice(["yes", "no", "limited"])
        seating = random.randint(10, 60)
        ambiance = random.choice(ambiances)
        google_rating = round(random.uniform(3.5, 4.9), 1)
        review_count = random.randint(20, 400)
    
    # Assign neighborhood
    neighborhood = neighborhoods[i % len(neighborhoods)]
    
    # Generate coordinates
    latitude = round(random.uniform(lat_min, lat_max), 6)
    longitude = round(random.uniform(lng_min, lng_max), 6)
    
    # Generate address
    street_num = random.randint(100, 9999)
    street_names = ["Jasper Ave", "Whyte Ave", "124 St", "104 St", "82 Ave", "Gateway Blvd", "Calgary Trail"]
    address = f"{street_num} {random.choice(street_names)}, Edmonton, AB"
    
    # Phone and website
    phone = f"780-{random.randint(100, 999)}-{random.randint(1000, 9999)}"
    website = f"https://{cafe_name.lower().replace(' ', '').replace('-', '')[:20]}.com" if random.random() > 0.3 else None
    
    # Hours
    hours_weekday = "7:00 AM - 6:00 PM" if random.random() > 0.3 else "6:30 AM - 7:00 PM"
    hours_weekend = "8:00 AM - 5:00 PM" if random.random() > 0.3 else "8:00 AM - 6:00 PM"
    
    # Opening date
    year = random.randint(2010, 2024)
    month = random.randint(1, 12)
    date_opened = f"{year}-{month:02d}-01"
    
    # Instagram
    instagram = f"@{cafe_name.lower().replace(' ', '').replace('-', '')[:20]}" if random.random() > 0.2 else None
    
    # Parking
    parking = random.choice(parking_options)
    
    # Timestamps
    created_at = datetime.now().isoformat()
    updated_at = created_at
    
    cafes_data.append((
        cafe_name, address, neighborhood, latitude, longitude, phone, website,
        cafe_type, ownership, avg_price, has_food, has_wifi, seating, ambiance,
        parking, hours_weekday, hours_weekend, date_opened, instagram,
        google_rating, review_count, created_at, updated_at
    ))

# Insert data
cursor.executemany("""
    INSERT INTO cafes (
        name, address, neighborhood, latitude, longitude, phone, website,
        cafe_type, ownership, avg_beverage_price, has_food, has_wifi,
        seating_capacity, ambiance, parking_availability, hours_weekday,
        hours_weekend, date_opened, instagram_handle, google_rating,
        review_count, created_at, updated_at
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
""", cafes_data)

conn.commit()

# Print summary
cursor.execute("SELECT COUNT(*) FROM cafes")
total_cafes = cursor.fetchone()[0]

cursor.execute("SELECT COUNT(*) FROM cafes WHERE name LIKE '%Square One%'")
soc_cafes = cursor.fetchone()[0]

print(f"✅ Synthetic data generated successfully!")
print(f"   Database: {db_path}")
print(f"   Total cafes: {total_cafes}")
print(f"   - Square One Coffee locations: {soc_cafes}")
print(f"   - Competitor cafes: {total_cafes - soc_cafes}")
print(f"   Neighborhoods covered: {len(neighborhoods)}")

conn.close()
